src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -61,6 +69,11 @@ internal sealed partial class ObjectReader : IDisposable
             _reader = new BinaryReader(stream, Encoding.UTF8);
             _objectReferenceMap = new ReaderReferenceMap<object>();
             _stringReferenceMap = new ReaderReferenceMap<string>();
+
+            // Capture a copy of the current static binder state.  That way we don't have to 
+            // access any locks while we're doing our processing.
+            _binderSnapshot = ObjectBinder.GetSnapshot();
```

Comment:
What is the guarantee that we don't rely on a typereader that gets added after this point?